pkg/apply/apply.go

```diff
@@ -118,9 +118,10 @@ func (changes *UpdateChangesTracker) mergeReplicaAssignments(
 
 // Union of NewChangesTracker and UpdateChangesTracker
 // used as a return type for the Apply function (which forks into applyNewTopic or applyExistingTopic)
-type NewOrUpdatedChanges struct {
-	NewChanges    *NewChangesTracker
-	UpdateChanges *UpdateChangesTracker
+type Changes struct {
+	NewChanges    *NewChangesTracker    `json:"newTopic"`
+	UpdateChanges *UpdateChangesTracker `json:"updatedTopic"`
+	DryRun        bool                  `json:"dryRun"`
```

Comment:
I don't think this data structure is the right representation for your result now.
Before you had a list of changes that were grouped into new topics and updated topics. 
That meant the Valid outputs could be a combination of both or two empty lists.

Now you are returning one topic changeset at a time.
That can only be an update or a new topic. It cannot be both.
This data structure allows it to be both thus leaving the door open to a return value that would be syntactically correct biut semantically invalid. This puts the burden of the verificaiton on the user of this data structure.

You can avoid this failure mode by refactoring the data structure to contain either a NewChanges or an UpdateChanges (Union) so it would be impossible to create an invalid structure and the user would only have valid combinations in output.

Reply:
What about making an interface with one method (`format_json`) and have two implementations ? One for the new and one for the updated topic ?
https://golangbot.com/polymorphism/